File: client/src/boundary/interactor/StartGameInteractor.ts
import { StartGameUC } from "boundary/api/StartGameUC";
import { GameD2BConverter } from "./GameD2BConverter";
import GameStorage from "services/api/GameStorage";
import BoundaryGame from "boundary/model/BoundaryGame";
import { map, Observable } from "rxjs";

export class StartGameInteractor implements StartGameUC {
  private gameStorage: GameStorage;
  private gameConverter: GameD2BConverter;

  constructor(gameStorage: GameStorage, gameConverter: GameD2BConverter) {
    this.gameStorage = gameStorage;
    this.gameConverter = gameConverter;
  }

  startGame(): Observable<BoundaryGame> {
    return this.gameStorage
      .startGame()
      .pipe(map((game) => this.gameConverter.convert(game)));
  }
}

File: server/battleships/use_cases/implementation/src/main/java/eu/optas/use_cases/implementation/StartGameInteractor.java
package eu.optas.use_cases.implementation;

import eu.optas.gateway.api.GameGateway;
import eu.optas.use_cases.api.BoundaryGame;
import eu.optas.use_cases.api.StartGameUC;

public class StartGameInteractor implements StartGameUC {
    private final GameGateway gameGateway;
    private final GameD2BConverter gameD2BConverter;

    public StartGameInteractor(GameGateway gameGateway, GameD2BConverter gameD2BConverter) {
        this.gameGateway = gameGateway;
        this.gameD2BConverter = gameD2BConverter;
    }

    @Override
    public BoundaryGame startGame() {
        return gameD2BConverter.convert(gameGateway.createGame());
    }
}
